packages/realm-react/src/RealmProvider.tsx

```diff
@@ -161,6 +193,47 @@ export function createRealmProvider(
   };
 }
 
+/**
+ * Generates a `RealmProvider` which is either based on a configuration
+ * or based on a realm, depending on its props.
+ * @param RealmContext - The context that will contain the Realm instance
+ * @returns a RealmProvider component that provides context to all context hooks
+ */
+export function createDynamicRealmProvider(RealmContext: React.Context<Realm | null>): DynamicRealmProvider {
+  return ({ realm, children, ...configurationProps }) => {
+    if (realm) {
+      if (!isEqual(configurationProps, {})) {
+        throw new Error("Cannot use configuration props when using an existing Realm instance.");
+      }
+
+      const RealmProviderFromRealm = createRealmProviderFromRealm(realm, RealmContext);
+      return <RealmProviderFromRealm>{children}</RealmProviderFromRealm>;
+    } else {
+      const RealmProviderFromConfig = createRealmProviderFromConfig({}, RealmContext);
+      return <RealmProviderFromConfig {...configurationProps}>{children}</RealmProviderFromConfig>;
```

Comment:
NIT: Since these are never in scope together, it could be simpler to just name them `RealmProvider` here.

```suggestion
      const RealmProvider = createRealmProviderFromRealm(realm, RealmContext);
      return <RealmProvider>{children}</RealmProvider>;
    } else {
      const RealmProvider = createRealmProviderFromConfig({}, RealmContext);
      return <RealmProvider {...configurationProps}>{children}</RealmProvider>;
```